### PR TITLE
cephfs: support WORM(Write Once Read Many) feature

### PR DESCRIPTION
**Pull request by weiqiaomiao**
Write Once Read Many (WORM) technology is used in many applications for data integrity reasons and because of the accepted legal admissibility of stored files that use the technology. And we can see some distribute filesystem such as glusterfs support WORM feature.  Driven by the needs of our customer ,  we are working on Cephfs support WORM(write once read only) feature.
      In our implementation, the worm feature is disabled default, and can enabled on a exist empty dir . A dir which enable worm feature called a worm root dir. When a subdir or a file create under worm root dir, it  auto inherit the worm attibute from worm root dir. A file which enable worm feature called a worm file. The worm file has three state：
**(1)  normal state：** the  initial state of a file. in this state, the worm file is mutable and deletable。
  **(2)  retained state:** in this state, the worm file is immutable or undeletable.
**(3)  expired state:**  in this state,  the worm like as the normal file, it is mutable and deletable (the worm file is immutable but deletable in some other storage).

  A file makes transition between these 3 states. Methods of making WORM/Retention transition:
**(1) manual commit:** Using chattr +i command to make the worm file through into retained state immediately.
**(2) automatic  commit:** Automatic transition between the different states of a file. Dormant Files (files which are untouched for a long time) will be converted into WORM files based on auto-commit period . 

Retention Profiles/Policies: Configurable Policies that guide the WORM/Retention behavior. Contains,
     **(1)  Default Retention Period**: default time period till which a WORM file should be retained(undeletable)
     **(2)  Auto-commit Period**:  a worm file get into retained state if it is untouched in  auto commit period.
     **(3)  minimum retention period:** minimum time period for users config on  a worm file
     **(4)   max retention period:** max time period for user config on a worm file 

**Cephfs Modification:**
      (1)  add  worm state and retention policies into inode per file
      (2)  worm feature can enable on any empty dir, but once it enabled, it can't disabled. we can enable worm feature by setfattr like:
>   setfattr -n ceph.worm -v "enable=1 auto_commit_period=120 min_retention_period=240 max_retention_period=7200 retention_period=360" dir1

  or
>    setfattr -n ceph.worm.enable -v 1  dir1
>    setfattr -n ceph.worm.enable -v 120  dir1
>    setfattr -n ceph.worm.min_retention_period -v 240  dir1
>    setfattr -n ceph.worm.max_retention_period -v 7200  dir1
>    setfattr -n ceph.worm.retention_period -v 360  dir1

and get worm state and policies by getfattr like:

>      getfattr -n ceph.worm dir1

             
  (3)   Lazy Auto-commit:  IO Triggered Using timeouts for untouched files. The modify  IO（write、truncate、unlink.etc) will cause the transition.

 (4)   Using the file atime save WORM file protection expiration time.
         **manual commit :** atime = mtime + retention_period
         **atomatic commit:** atime = mtime + auto_commit_period + retention_period
         user can change the worm file retention period ( must between minimum retention period ~ max retention period) through change file Atime. 

  (5)   hard link don't support if worm feature enable.

Signed-off-by:  Wei Qiaomiao wei.qiaomiao@zte.com.cn

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/common/options.cc [hunk 1]
```diff
@@ -7881,6 +7881,26 @@ std::vector<Option> get_mds_options() {
...
+    
+    Option("mds_request_load_average_decay_rate", Option::TYPE_FLOAT, Option::LEVEL_ADVANCED)
+    .set_default(60)
+    .set_description("rate of decay in seconds for calculating request load average"),
+
+    Option("mds_worm_commit_period", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
+     .set_default(120)
+     .set_description("default auto commit period for worm feature"),
+
+    Option("mds_worm_retention_period", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
```
Comment (xiexingguo) [inline src/common/options.cc]: Seems these three options are all sharing the same default value. Did you do that deliberately?

Also I've observed you silently changed the perm of all files you touched from 100644->100755, which I guess should definitely need to be addressed first..

Comment (weiqiaomiao) [reply src/common/options.cc]: These options had been redesigned， I think `mds_worm_retention_period`  may be equal to `mds_worm_min_retention_period`  and set them to  "1h",  `mds_worm_max_retention_period`  may be larger than 'mds_worm_min_retention_period` and set it to "30 years", What do you think?

the modified file mode already restore to 100644,  thanks!

File: src/common/options.cc [hunk 2]
```diff
@@ -7881,6 +7881,26 @@ std::vector<Option> get_mds_options() {
      .set_default(0)
      .set_description("threshold for cache usage to disallow \"dump cache\" operation to file")
      .set_long_description("Disallow MDS from dumping caches to file via \"dump cache\" command if cache usage exceeds this threshold."),
+    
+    Option("mds_request_load_average_decay_rate", Option::TYPE_FLOAT, Option::LEVEL_ADVANCED)
+    .set_default(60)
+    .set_description("rate of decay in seconds for calculating request load average"),
+
+    Option("mds_worm_commit_period", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
+     .set_default(7200)
```
Comment (xiexingguo) [inline src/common/options.cc]: s/7200/2_hr/

File: src/common/options.cc [hunk 3]
```diff
@@ -7881,6 +7881,26 @@ std::vector<Option> get_mds_options() {
...
+     .set_description("default auto commit period for worm feature"),
+
+    Option("mds_worm_retention_period", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
+     .set_default(3600)
```
Comment (xiexingguo) [inline src/common/options.cc]: s/3600/1_hr/

File: src/common/options.cc (hunk 1 above)
Comment (xiexingguo) [reply src/common/options.cc]: You might switch to the inline .set_min & .set_max helpers to perform the boundary checking, then

File: src/common/legacy_config_opts.h [hunk 4]
```diff
@@ -515,6 +515,12 @@ OPTION(mds_damage_table_max_entries, OPT_INT)
 // Maximum increment for client writable range, counted by number of objects
 OPTION(mds_client_writeable_range_max_inc_objs, OPT_U32)
 
+//worm attribute default value
+OPTION(mds_worm_commit_period, OPT_U64)
+OPTION(mds_worm_retention_period, OPT_U64)
+OPTION(mds_worm_min_retention_period, OPT_U64)
+OPTION(mds_worm_max_retention_period, OPT_U64)
```
Comment (ukernel) [inline src/common/legacy_config_opts.h]: no need to add options to  legacy_config_opts.h

File: src/mds/mdstypes.h [hunk 5]
```diff
@@ -377,6 +384,81 @@ inline std::ostream& operator<<(std::ostream &out, const vinodeno_t &vino) {
...
+    ENCODE_START(1, 1, bl);
+    encode(worm_state, bl);
+    encode(auto_commit_period, bl);
+    encode(retention_period, bl);
+    encode(min_retention_period, bl);
+    encode(max_retention_period, bl);
+    ENCODE_FINISH(bl);
+  }
+  void decode(bufferlist::iterator& p) {
+    DECODE_START_LEGACY_COMPAT_LEN(1, 1, 1, p);
```
Comment (ukernel) [inline src/mds/mdstypes.h]: DECODE_START()

File: src/mds/CInode.cc [hunk 6]
```diff
@@ -3687,6 +3687,9 @@ int CInode::encode_inodestat(bufferlist& bl, Session *session,
     }
   }
 
+  mempool_inode *policy_i = ppolicy ? pi : oi;
+  ::encode(policy_i->worm, bl);
+
```
Comment (ukernel) [inline src/mds/CInode.cc]: needs to bump encoding version

File: src/messages/MClientReply.h [hunk 7]
```diff
@@ -254,6 +256,8 @@ struct InodeStat {
         change_attr = 0;
       }
     }
+    
+    ::decode(worm, p);
```
Comment (ukernel) [inline src/messages/MClientReply.h]: needs to check encoding version

File: src/common/options.cc (hunk 1 above)
Comment (weiqiaomiao) [reply src/common/options.cc]: fixed it，thanks

File: src/common/options.cc (hunk 2 above)
Comment (weiqiaomiao) [reply src/common/options.cc]: fixed it，thanks

File: src/common/options.cc (hunk 3 above)
Comment (weiqiaomiao) [reply src/common/options.cc]: fixed it，thanks

File: src/common/legacy_config_opts.h (hunk 4 above)
Comment (weiqiaomiao) [reply src/common/legacy_config_opts.h]: ok，fixed it，thanks

File: src/mds/mdstypes.h (hunk 5 above)
Comment (weiqiaomiao) [reply src/mds/mdstypes.h]: fixed it，thanks

File: src/mds/CInode.cc (hunk 6 above)
Comment (weiqiaomiao) [reply src/mds/CInode.cc]: fixed it，thanks

File: src/messages/MClientReply.h (hunk 7 above)
Comment (weiqiaomiao) [reply src/messages/MClientReply.h]: fixed it，thanks

**Comment by weiqiaomiao** [comment]
@ukernel @batrick @jtlayton  I reorganized all the commit, and makes the code more readable now, can you take a look it？

File: src/client/Client.cc [hunk 8]
```diff
@@ -7041,6 +7037,34 @@ int Client::_do_setattr(Inode *in, struct ceph_statx *stx, int mask,
     req->inode_drop |= CEPH_CAP_FILE_CACHE | CEPH_CAP_FILE_RD |
       CEPH_CAP_FILE_WR;
   }
+
+  if (mask & CEPH_SETATTR_MODE) {
+    req->head.args.setattr.mode = stx->stx_mode;
+    req->inode_drop |= CEPH_CAP_AUTH_SHARED;
+     
+    if (in->worm.is_enable() && !in->worm.is_retain() && !(stx->stx_mode & 00222)) {
```
Comment (jtlayton) [inline src/client/Client.cc]: I'm not sure I really like treating a mode with no write bits as special.

What about POSIX ACLs here? Suppose there are no write bits in the mode but there is a POSIX ACL set that explicitly allows write access for another user? Should that trigger this transition?

A better mechanism might be to wire up the appropriate ioctl so that you could use "chattr -i" here. That would make it a privileged operation, but that seems reasonable for this.

**Comment by renhwztetecs** [comment]
As a document backup application scenario, WORM is an important function. It can protect files from being deleted and modified easily.

File: src/client/Client.cc (hunk 8 above)
Comment (weiqiaomiao) [reply src/client/Client.cc]: Thanks for your suggestion, jtlaytion!

My original intention was to find a simple way to trigger the conversion of the state and removing the Write property in mode was just one of the methods, and I didn't take into account the POSIX ACL scene. 

 I think your suggestion is more reasonable, but I'm not familiar with chattr,  so I think  I need a little time to learn how to do.

File: src/client/Client.cc [hunk 9]
```diff
@@ -6934,8 +6979,27 @@ int Client::_do_setattr(Inode *in, struct ceph_statx *stx, int mask,
     if (mask & (CEPH_SETATTR_MTIME|CEPH_SETATTR_ATIME)) {
       if (mask & CEPH_SETATTR_MTIME)
         in->mtime = utime_t(stx->stx_mtime);
-      if (mask & CEPH_SETATTR_ATIME)
-        in->atime = utime_t(stx->stx_atime);
+      if (mask & CEPH_SETATTR_ATIME) {
+        uint32_t min_atime_sec = in->mtime.sec()+ in->worm.min_retention_period;
```
Comment (renhwztetecs) [inline src/client/Client.cc]: Since the acquisition time comes from the local, there is a modification of the local time to cause the folder to enter the protection period in advance. It may be provided by providing an external standard clock source interface or in the document: "Synchronize the time before setting the WORM function"

What do you think? @ukernel @batrick @jtlayton

File: src/client/Client.cc (hunk 8 above)
Comment (jtlayton) [reply src/client/Client.cc]: Basically, you'd just need to wire up FS_IOC_GETFLAGS and FS_IOC_SETFLAGS and fix them up to handle the IMMUTABLE flag. FUSE has an ioctl handler, so it should be possible to do this for both the kernel and fuse.

File: src/client/Client.cc (hunk 9 above)
Comment (jtlayton) [reply src/client/Client.cc]: I think it'd be better to not overload the atime field. While it's not terribly common, some applications _do_ use it. Why not store this in a separate field in the inode and present it via xattr?

File: src/client/Client.cc (hunk 8 above)
Comment (weiqiaomiao) [reply src/client/Client.cc]: I'm very sorry for delay in reply.  I  add ll_set_flags() function to handle IMMUTABLE flag (make 
 the worm state transition in it) and ll_get_flags() function to get  the flags . Can you take a look at it?

File: src/client/Client.cc (hunk 9 above)
Comment (weiqiaomiao) [reply src/client/Client.cc]: OK. Already drop to use atime and add a new field in worm info struct insteed.

**Comment by batrick** [comment]
This looks interesting but before I dig into it I'd like you to cleanup the commit history. For example, 58fd92acac65c19f35418dfbcc9230681b4f04e8 needs rewritten to remove the addition to `src/common/legacy_config_opts.h`. Use `git rebase -i` and edit the commits as necessary. Please also make sure you don't add unnecessary whitespace changes. I also noticed a lot of trailing whitespace in your patches.

**Comment by weiqiaomiao** [comment]
jenkins，retest this please

File: src/client/Client.cc [hunk 10]
```diff
@@ -8736,14 +8744,21 @@ int Client::_open(Inode *in, int flags, mode_t mode, Fh **fhp,
     return -EROFS;
   }
 
+  int result = 0;
+  if (in->is_file() && (flags & (O_WRONLY | O_RDWR  | O_TRUNC | O_APPEND))&& in->worm.is_enable()) {
```
Comment (jtlayton) [inline src/client/Client.cc]: nit: please fix the whitespace above

File: src/mds/Server.h [hunk 11]
```diff
@@ -347,6 +347,8 @@ class Server {
   void handle_conf_change(const ConfigProxy& conf,
                           const std::set <std::string> &changed);
 
+ // int worm_state_transition(Inode *in, int op);
+
```
Comment (jtlayton) [inline src/mds/Server.h]: This doesn't look related to the rest of the patch.

File: src/client/Client.cc [hunk 12]
```diff
@@ -11690,6 +11705,62 @@ int Client::ll_removexattr(Inode *in, const char *name, const UserPerm& perms)
   return _removexattr(in, name, perms);
 }
 
+int Client::worm_state_transition(Inode *in, const UserPerm& perms, int op)
```
Comment (jtlayton) [inline src/client/Client.cc]: A comment header would be nice here. What, specifically, is this function intended to do?

File: src/client/fuse_ll.cc [hunk 13]
```diff
@@ -651,6 +651,44 @@ static void fuse_ll_ioctl(fuse_req_t req, fuse_ino_t ino, int cmd, void *arg, st
...
+       fuse_reply_ioctl(req, 0, &attr, sizeof(attr));
+    }
+    break;
+    case FS_IOC_SETFLAGS: {
+      auto flags = (int *)in_buf;
+      
+      if (*flags & (~FS_IMMUTABLE_FL)) {
+        fuse_reply_err(req, EOPNOTSUPP);
+        return;
+      }
```
Comment (jtlayton) [inline src/client/fuse_ll.cc]: I'm a little concerned with this. Suppose I accidentally set this on the wrong file. How do I fix it?

File: src/client/Client.cc [hunk 14]
```diff
@@ -13026,6 +13026,11 @@ int Client::_link(Inode *in, Inode *dir, const char *newname, const UserPerm& pe
     return -EDQUOT;
   }
 
+  if (dir->worm.is_enable() || in->worm.is_enable()) {
+    ldout(cct, 8) << "_link" << "worm feature doesn't support link " << dendl;
+    return -EPERM;
+  }
```
Comment (jtlayton) [inline src/client/Client.cc]: Why not?

File: src/client/Client.cc (hunk 10 above)
Comment (weiqiaomiao) [reply src/client/Client.cc]: fixed it，thanks

File: src/mds/Server.h (hunk 11 above)
Comment (weiqiaomiao) [reply src/mds/Server.h]: already deleted

File: src/client/Client.cc (hunk 12 above)
Comment (weiqiaomiao) [reply src/client/Client.cc]: OK

File: src/client/fuse_ll.cc (hunk 13 above)
Comment (weiqiaomiao) [reply src/client/fuse_ll.cc]: I think may be only wait for file protection time expires 。any suggestion？

Comment (jtlayton) [reply src/client/fuse_ll.cc]: I don't think waiting will be acceptable -- we will need to be able to fix this in some fashion, as we can't just throw away the filesystem and start over (as you would with a CD-R or something).

Maybe this could only be allowed from certain clients using a new cephfs client capability flag, and those clients would be able to undo this as well?

http://docs.ceph.com/docs/master/cephfs/client-auth/

File: src/mds/Server.cc [hunk 15]
```diff
@@ -5135,6 +5149,123 @@ void Server::create_quota_realm(CInode *in)
...
+    } else if (name == "worm.retention_period") {
+      uint32_t q = boost::lexical_cast<uint32_t>(value);
+      if (worm->is_enable() && !worm->is_root())
+        return -EINVAL;
+
+      if (!worm->is_enable()){
+         parse_worm_vxattr(mdr, cur, "worm.enable", "1", worm);
+      }
+
+      if (q == 0xffffffff) {                    // retetion period ulimit
```
[the remark this answers is not in the document]
Comment (weiqiaomiao) [reply src/mds/Server.cc]: OK

File: src/mds/mdstypes.h [hunk 16]
```diff
@@ -73,6 +73,13 @@
 #define MDS_TRAVERSE_DISCOVER      2    // skips permissions checks etc.
 #define MDS_TRAVERSE_DISCOVERXLOCK 3    // succeeds on (foreign?) null, xlocked dentries.
 
+#define WORM_ENABLE  1     // worm feature enbale
+#define WORM_IS_ROOT 2     // worm root dir
+#define WORM_RETAIN  4     // the worm file is in retained state
+#define WORM_EXPIRE  8     // the worm file is in expired state
+#define WORM_RETEN_PERIOD_UNLIMIT 16  //the worm file retention period is unlimit
+#define MAX_MIN_RETEN_PERIOD 946080000 // 30 years
+#define MIN_MAX_RETEN_PERIOD 2207520000 // 70 years
```
[the remark this answers is not in the document]
Comment (weiqiaomiao) [reply src/mds/mdstypes.h]: Our customers require a minimum retention period that can be set from 0 to 30 years, and a maximum retention period from 0 to 70 years. And to avoid other people being confused like you, I will remove them and put the restrictions on the configuration of our own management system.

File: src/client/Client.cc (hunk 14 above)
Comment (weiqiaomiao) [reply src/client/Client.cc]: We expect all files in the worm directory to be controlled，but it hardlink is supported，if a non-worm file linked to a file under worm directory, it can't  controlled by worm.

File: src/client/fuse_ll.cc (hunk 13 above)
Comment (weiqiaomiao) [reply src/client/fuse_ll.cc]: OK ! Already add a new cap flag "i" and "chattr -i"" to expired a file which is in retained state.

**Comment by weiqiaomiao** [comment]
jenkins retest this please
